### ISSUE TITLE
Harmonising wildcard processing

### ISSUE THREAD

**Issue by olejandro** [etsap-TIMES/xl2times#209]
Currently there are 2 transforms dealing with wildcard processing:
- `process_uc_wildcards` deals with wildcards in `uc_t` tables;
- `process_wildcards` deals with wildcards in `tfm` tables and changes the dataframe with attribute data based on their content.

Would it be practical to:
- separate the modification logic of the dataframe with attribute data based on the content of `tfm` tables out of `process_wildcards`;
- combine wildcard processing done in `process_uc_wildcards` and `process_wildcards` in a single transform?

I believe these changes would also allow addressing #153 and #154 easier.

**Comment by olejandro** [comment]
@siddharth-krishna @SamRWest what do you think?

**Comment by siddharth-krishna** [comment]
The only potential issue I can see with separating e.g. `process_wildcards` into two steps, one that expands the wildcards and the second to apply/insert/update the resulting rows, is if the intermediate table with the expanded rows is huge and consumes a lot of memory. Right now, we expand the wildcards in each row and immediately apply/insert/update it, IIRC, which avoids the creation of this intermediate expanded table. If you think it's not likely for intermediate tables to be too large, then your plan sounds good.

An alternative idea might be to move the helper functions inside `process_wildcards` to `utils` or something and also use them in `process_uc_wildcards`. This would avoid creation of such intermediate tables. My intent in #166 was anyway that the helper functions should eventually live somewhere more general (and perhaps even in the future allow a jupyter-based scenario workflow).

**Comment by olejandro** [comment]
Memory can definitely be an issue. Could we e.g., store resolved wildcards as a list / string in in the original table until it is used or would that not help?

**Comment by SamRWest** [comment]
The fastest approach I've found (and what I've implemented in `_match_uc_wildcards()`) is to essentially pre-build a lookup table with all unique wildcard patterns and their list of matches, then join these back to the table (thus re-duplicating them) and explode the match-lists for each row (making the DF long format).

The regex lookups are slow, and the DF row iteration is incredibly slow.  This minimises the regex lookups and avoid the iteration entirely, making it much faster.

There are probably some additional gains to be had by building this lookup table of unique wildcard matches once for wildcards in _all_ tables, as this should eliminate additional duplicates.  Yes it'll use more RAM to store, but it'll be far less than the end-result joined/exploded tables, so I doubt it'd a bottleneck.

I've got this working for ~TFM_UPD [over here](https://github.com/etsap-TIMES/xl2times/commit/1c2c08bf1fe3eeb621db478413382f2c98176edf#diff-6c555b05eed75bd395d795a00453b016b080d9dd690d46dd2c1dd6d65b5beb3eR2326) but got sidetracked on some other stuff and need to get back to it.

I've opened PR #210 for you guys to have a look at.  Happy to push on if you think this is the right approach?
